### PR TITLE
Add z-loss to Bamba for v2

### PR DESCRIPTION
**Pull request by daviswer**
# What does this PR do?

Adds support for auxiliary z-loss when tuning Bamba v2. Also fixes some typos in the checkpoint conversion utility script. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@fabianlim @ ArthurZucker ?

### CODE REVIEW

File: src/transformers/models/bamba/convert_mamba_ssm_checkpoint.py [hunk 1]
```diff
@@ -248,7 +248,7 @@ def convert_mamba_ssm_checkpoint_file_to_huggingface_model_file(
         "--precision",
         type=str,
         default="fp16",
-        const="fp16",
+        # const="fp16",
```
Comment (fabianlim) [inline src/transformers/models/bamba/convert_mamba_ssm_checkpoint.py:251]: this is not used anywhere?

File: src/transformers/models/bamba/convert_mamba_ssm_checkpoint.py [hunk 2]
```diff
@@ -267,7 +267,8 @@ def convert_mamba_ssm_checkpoint_file_to_huggingface_model_file(
     args = parser.parse_args()
 
     convert_mamba_ssm_checkpoint_file_to_huggingface_model_file(
-        args.mamba2_checkpoint_directory,
+        args.mamba_ssm_checkpoint_directory,
```
Comment (fabianlim) [inline src/transformers/models/bamba/convert_mamba_ssm_checkpoint.py:270]: was this a typo?

File: src/transformers/models/bamba/modeling_bamba.py [hunk 3]
```diff
@@ -1521,6 +1522,8 @@ def forward(
         loss = None
         if labels is not None:
             loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size, **kwargs)
```
Comment (fabianlim) [inline src/transformers/models/bamba/modeling_bamba.py:1524]: this file has to be generated from `modular_bamba`

File: src/transformers/models/bamba/convert_mamba_ssm_checkpoint.py (hunk 2 above)
Comment (daviswer) [reply src/transformers/models/bamba/convert_mamba_ssm_checkpoint.py:270]: Yes

File: src/transformers/models/bamba/modeling_bamba.py (hunk 3 above)
Comment (fabianlim) [reply src/transformers/models/bamba/modeling_bamba.py:1524]: `python utils/modular_model_converter.py --files_to_parse src/transformers/models/bamba/modular_bamba.py`

Comment (daviswer) [reply src/transformers/models/bamba/modeling_bamba.py:1524]: Never authored a modular HF file before... how does one do this?

File: src/transformers/models/bamba/modular_bamba.py [hunk 4]
```diff
@@ -1231,21 +1241,44 @@ def forward(
...
+            cache_position=cache_position,
             **kwargs,
         )
 
+        hidden_states = outputs.last_hidden_state
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
+        logits = self.lm_head(hidden_states[:, slice_indices, :])
+
+        loss = None
```
Comment (fabianlim) [inline src/transformers/models/bamba/modular_bamba.py:1268]: can we just do this? if so then we dont need so many changes.. also best to assign .0001 to some variable

```
outputs = super().forward(...)
outputs.loss += .0001 * logits.logsumexp(dim=-1).pow(2).mean()
```

File: src/transformers/models/bamba/modular_bamba.py [hunk 5]
```diff
@@ -1185,6 +1185,16 @@ class BambaForCausalLM(LlamaForCausalLM):
     @can_return_tuple
     @add_start_docstrings_to_model_forward(BAMBA_INPUTS_DOCSTRING)
     @replace_return_docstrings(output_type=CausalLMOutputWithPast, config_class=_CONFIG_FOR_DOC)
+    def __init__(self, config):
+        super().__init__(config)
+        self.model = BambaModel(config)
+        self.vocab_size = config.vocab_size
+        self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)
+        self.use_z_loss = config.use_z_loss
```
Comment (vasqu) [inline src/transformers/models/bamba/modular_bamba.py:1193]: Only `self.use_z_loss` is new, no? Do we need all of the other stuff after the `super` call?

File: src/transformers/models/bamba/modular_bamba.py [hunk 6]
```diff
@@ -1231,21 +1241,44 @@ def forward(
...
+        hidden_states = outputs.last_hidden_state
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
+        logits = self.lm_head(hidden_states[:, slice_indices, :])
+
+        loss = None
+        if labels is not None:
+            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size, **kwargs)
+            if self.use_z_loss:
+                loss = loss + .0001 * logits.logsumexp(dim=-1).pow(2).mean()
```
Comment (vasqu) [inline src/transformers/models/bamba/modular_bamba.py:1272]: Should the `.0001` be another new (config) attribute maybe?

Also, could you check if the dtypes are correct here: the loss might be `float32` while logits could be half for example. Might need an upcast here.

File: src/transformers/models/bamba/modular_bamba.py [hunk 7]
```diff
@@ -1185,6 +1185,16 @@ class BambaForCausalLM(LlamaForCausalLM):
     @can_return_tuple
     @add_start_docstrings_to_model_forward(BAMBA_INPUTS_DOCSTRING)
     @replace_return_docstrings(output_type=CausalLMOutputWithPast, config_class=_CONFIG_FOR_DOC)
```
Comment (vasqu) [inline src/transformers/models/bamba/modular_bamba.py:1187]: Seems like the decorators got unintentionally moved 👀 (they belong to the forward)

**Review by vasqu — COMMENTED** [review]
Left a few comments. But shouldn't be too much / difficult to fix.

File: src/transformers/models/bamba/modular_bamba.py (hunk 4 above)
Comment (vasqu) [reply src/transformers/models/bamba/modular_bamba.py:1268]: Not sure if modular can do this but worth a try :D

File: src/transformers/models/bamba/modular_bamba.py (hunk 6 above)
Comment (fabianlim) [reply src/transformers/models/bamba/modular_bamba.py:1272]: maybe instead of `use_z_loss`, you might want to say `z_loss_coef` and have it default to `None`, which will turn the zloss off in that case.

Comment (vasqu) [reply src/transformers/models/bamba/modular_bamba.py:1272]: Also a fair approach. I don't think either is  vastly superior so I'd leave it up to your preferences.

File: src/transformers/models/bamba/modeling_bamba.py (hunk 3 above)
Comment (daviswer) [reply src/transformers/models/bamba/modeling_bamba.py:1524]: Great, I'll do that. Thanks!

File: src/transformers/models/bamba/modular_bamba.py (hunk 5 above)
Comment (daviswer) [reply src/transformers/models/bamba/modular_bamba.py:1193]: I pulled that from the `modeling_` code, wasn't sure if `modular` was handling the bamba-specific stuff. Looks like it is though, so I'll clean that up

File: src/transformers/models/bamba/modular_bamba.py (hunk 7 above)
Comment (daviswer) [reply src/transformers/models/bamba/modular_bamba.py:1187]: oops, good catch

File: src/transformers/models/bamba/convert_mamba_ssm_checkpoint.py [hunk 8]
```diff
@@ -267,7 +267,8 @@ def convert_mamba_ssm_checkpoint_file_to_huggingface_model_file(
...
         args.precision,
         args.output_dir,
+        save_model = "sharded",
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/models/bamba/convert_mamba_ssm_checkpoint.py:273]: Fair point. I was just wondering since the default value (`True`) indicated a single shard (but that just might've been a wrong default value from the beginning). 

Imo, we can keep this change since bamba models usually have multiple shards iirc.

File: src/transformers/models/bamba/modular_bamba.py (hunk 4 above)
Comment (daviswer) [reply src/transformers/models/bamba/modular_bamba.py]: Looks like modular didn't take it. Switching back...

File: src/transformers/models/bamba/configuration_bamba.py [hunk 9]
```diff
@@ -100,6 +100,8 @@ class BambaConfig(PretrainedConfig):
             Flag indicating whether or not to use bias in the convolution layer of the mamba mixer block.
         mamba_proj_bias (`bool`, *optional*, defaults to `False`):
             Flag indicating whether or not to use bias in the input and output projections (["in_proj", "out_proj"]) of the mamba mixer block
+        zloss_coeff (`float`, *optional*, defaults to 0):
```
Comment (vasqu) [inline src/transformers/models/bamba/configuration_bamba.py:103]:
```suggestion
        z_loss_coeff (`float`, *optional*, defaults to 0):
```
More of a nit.

Comment (vasqu) [reply src/transformers/models/bamba/configuration_bamba.py:103]: Needs to be changed elsewhere then.

File: src/transformers/models/bamba/modular_bamba.py [hunk 10]
```diff
@@ -1231,21 +1238,44 @@ def forward(
...
+        hidden_states = outputs.last_hidden_state
+        # Only compute necessary logits, and do not upcast them to float if we are not computing the loss
+        slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
+        logits = self.lm_head(hidden_states[:, slice_indices, :])
+
+        loss = None
+        if labels is not None:
+            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size, **kwargs)
+            if self.zloss_coeff > 0:
+                loss = loss + self.zloss_coeff * logits.logsumexp(dim=-1).to(dtype=loss.dtype).pow(2).mean()
```
Comment (vasqu) [inline src/transformers/models/bamba/modular_bamba.py:1269]:
```suggestion
                loss = loss + self.zloss_coeff * logits.to(dtype=loss.dtype).logsumexp(dim=-1).pow(2).mean()
```
Wouldn't we typecast before applying the ops?

**Review by vasqu — COMMENTED** [review]
Just another round of smaller things. Overall, looks good tho.

File: src/transformers/models/bamba/modular_bamba.py (hunk 10 above)
Comment (daviswer) [reply src/transformers/models/bamba/modular_bamba.py:1269]: In the case where loss is higher precision than logits, we'd want to avoid upcasting the massive logits tensor until after reducing on one of the dimensions

Comment (daviswer) [reply src/transformers/models/bamba/modular_bamba.py]: Added a comment to this effect

File: src/transformers/models/bamba/configuration_bamba.py [hunk 11]
```diff
@@ -186,6 +189,7 @@ def __init__(
         self.mamba_chunk_size = mamba_chunk_size
         self.mamba_conv_bias = mamba_conv_bias
         self.mamba_proj_bias = mamba_proj_bias
+        self.zloss_coeff = z_loss_coefficient
```
Comment (vasqu) [inline src/transformers/models/bamba/configuration_bamba.py:192]:
```suggestion
        self.z_loss_coefficient = z_loss_coefficient
```
Let's propogate it everywhere 👀 (also in modular file etc)

**Comment by vasqu** [comment]
The ci failures can be fixed with `make style` :)

File: src/transformers/models/bamba/configuration_bamba.py (hunk 11 above)
Comment (daviswer) [reply src/transformers/models/bamba/configuration_bamba.py:192]: ah ok

**Comment by vasqu** [comment]
Hopefully last fix needed: `make fix-copies`

File: src/transformers/models/bamba/configuration_bamba.py [hunk 12]
```diff
@@ -135,6 +137,7 @@ def __init__(
         mamba_chunk_size=256,
         mamba_conv_bias=True,
         mamba_proj_bias=False,
+        z_loss_coefficient=False,
```
Comment (vasqu) [inline src/transformers/models/bamba/configuration_bamba.py:140]:
```suggestion
        z_loss_coefficient=0.0,
```

File: src/transformers/models/bamba/configuration_bamba.py [hunk 13]
```diff
@@ -100,6 +100,8 @@ class BambaConfig(PretrainedConfig):
             Flag indicating whether or not to use bias in the convolution layer of the mamba mixer block.
         mamba_proj_bias (`bool`, *optional*, defaults to `False`):
             Flag indicating whether or not to use bias in the input and output projections (["in_proj", "out_proj"]) of the mamba mixer block
+        z_loss_coefficient (`float`, *optional*, defaults to 0):
```
Comment (vasqu) [inline src/transformers/models/bamba/configuration_bamba.py:103]:
```suggestion
        z_loss_coefficient (`float`, *optional*, defaults to 0.0):
```

File: src/transformers/models/bamba/configuration_bamba.py (hunk 12 above)
Comment (vasqu) [reply src/transformers/models/bamba/configuration_bamba.py:140]: Pretty sure that's what the ci was complaining about since default values differed between docs and actual value.

**Comment by vasqu** [comment]
Thanks, lgtm! Can you put it out of draft? 

cc @ArthurZucker @Cyrilvallez for core maintainer

**Review by Cyrilvallez — APPROVED** [review]
Hey! Super sorry for being late on this! LGTM, merging! 🤗

**Comment by fabianlim** [comment]
@Cyrilvallez hey! just checking in on this PR.. it sounded like it was going to be merged two weeks ago. Is there some other work required?

**Comment by Cyrilvallez** [comment]
Oh super sorry, I had enabled auto-merge, but it failed apparently 🤔 Not sure why 😅 Merging now after branch update, thanks again! 🤗
